src/torsion_profile.py

```diff
@@ -0,0 +1,191 @@
+        engine_args['freeze_atoms'] = self.freeze_atoms
+        self.engine = self.engine_(target=self, mol=self.mol, **engine_args)
+
+    def read_reference_data(self):
+
+        """ Read the reference ab initio data from a file such as qdata.txt.
+
+        After reading in the information from qdata.txt, it is converted
+        into the GROMACS energy units (kind of an arbitrary choice).
+        """
+        ## Reference (QM) energies
+        self.eqm           = []
+        ## The qdata.txt file that contains the QM energies and forces
+        self.qfnm = os.path.join(self.tgtdir,"qdata.txt")
+        # Parse the qdata.txt file
+        for line in open(os.path.join(self.root,self.qfnm)):
+            sline = line.split()
+            if len(sline) == 0: continue
+            elif sline[0] == 'ENERGY':
+                self.eqm.append(float(sline[1]))
+
+        if len(self.eqm) != self.ns:
+            raise RuntimeError("Length of qdata.txt should match number of structures")
+
+        # Turn everything into arrays, convert to kcal/mol
+        self.eqm = np.array(self.eqm)
+        self.eqm *= eqcgmx / 4.184
+        # Use the minimum energy structure of the QM as reference
+        self.eqm  -= np.min(self.eqm)
+        self.smin  = np.argmin(self.eqm)
+        logger.info("Referencing all energies to the snapshot %i (minimum energy structure in QM)\n" % self.smin)
+
+        if self.attenuate:
+            # Attenuate energies by an amount proportional to their
+            # value above the minimum.
+            eqm1 = self.eqm - np.min(self.eqm)
+            denom = self.energy_denom
+            upper = self.energy_upper
+            self.wts = np.ones(self.ns)
+            for i in range(self.ns):
+                if eqm1[i] > upper:
+                    self.wts[i] = 0.0
+                elif eqm1[i] < denom:
+                    self.wts[i] = 1.0 / denom
+                else:
+                    self.wts[i] = 1.0 / np.sqrt(denom**2 + (eqm1[i]-denom)**2)
+        else:
+            self.wts = np.ones(self.ns)
+
+        # Normalize weights.
+        self.wts /= sum(self.wts)
+
+    def indicate(self):
+        title_str = "Torsion Profile: %s, Objective = % .5e, Units = kcal/mol, Angstrom" % (self.name, self.objective)
+        #QYD: This title is carefully placed to align correctly
+        column_head_str1 =  "%-50s %-10s %-12s %-18s %-12s %-10s %-11s %-10s" % ("System", "Min(QM)", "Min(MM)", "Range(QM)", "Range(MM)", "Max-RMSD", "Ene-RMSE", "Obj-Fn")
+        printcool_dictionary(self.PrintDict,title=title_str + '\n' + column_head_str1, keywidth=50, center=[True,False])
+
+    def get(self, mvals, AGrad=False, AHess=False):
+        Answer = {'X':0.0, 'G':np.zeros(self.FF.np), 'H':np.zeros((self.FF.np, self.FF.np))}
+        self.PrintDict = OrderedDict()
+        # enable self.system_mval_masks (still need to implement)
+        enable_system_mval_mask = hasattr(self, 'system_mval_masks')
```

Comment:
These two lines are not needed here.

Reply:
Changes have been made in the latest commit.